Week3 [STEP 2] Hisop

Hi-sop opened the pull request:
@juun97 
STEP2 PR드립니다.
 
## 고민했던 점 
특별히 없음 

대부분의 요구사항을 step1때 구현하여 테스트코드 추가 및 테스트만 진행했습니다.

Commit:
Test: Step2에 해당하는 테스트 추가
대부분의 기능도 저번에 만들어서 추가 수정이 필요해보이지 않음

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -66,7 +66,11 @@ struct CoffeeShop {
 
 var misterLee = Person(name: "misterLee")
-var missKim = Person(name: "missKim")
+var missKim = Person(name: "missKim", money: 3000)
 
 var yagombucks = CoffeeShop()
 yagombucks.barista = misterLee
+
+missKim.buyCoffee(.americano, from: &yagombucks)
+missKim.money = 5000
```
missKim 인스턴스를 만들 때 initializer 를 사용해서 money 프로퍼티를 초기화 해주셨는데요,
하단에서 money 프로퍼티에 대해 직접 값을 변경해주고 있습니다.

특정 타입의 프로퍼티를 직접 수정하는 것은 일반적으로 권장되지 않는 방법입니다. 이는 객체 지향 프로그래밍의 캡슐화 원칙에 반하는 행동일 수 있으며, 코드의 유지보수성과 확장성을 저해할 수 있습니다.

프로퍼티의 수정은 해당 객체가 제공하는 메서드를 통해서 이루어져야 합니다.

juun97 reviewed (commented):
코드의 변경사항이 거의 없어서 저번 스텝에서 말씀드리지 않았던 사항을 코멘트로 남기겠습니다.


1. unwrapedMenu 

unwrapedMenu 의 네이밍에 대해서 고민을 해보시면 좋을 것 같습니다. 현재 네이밍은 작성한 개발자에게 더 친숙한 네이밍인 것 같습니다. 옵셔널 바인딩을 한 값이 어떤 값(Dictonary의 Value)인지 생각을 해보고 다시 네이밍을 해볼까요?


2. pickUPTable

- 새로운 값 추가

Hisop님은 현재 Dictionary의 값을 추가할 때 
```swift
pickUpTable[name]?.append(coffee)
```

위 와 같은방식을 사용하고 계십니다. 가운데 보면 `?` 표시가 붙게 되있죠. 왜냐하면 pickUPTable 에는 name 에 속하는 key 가 있을 수도 있고 없을 수도 있기 때문입니다. 현재 상황에서는 pickUPTable 에 coffee 가 저장이 되고 있을까요?

- 완료 메세지

makeCoffe 메서드가 호출한 순간 직접 print 를 해주고 계신데요. 현재의 경우 코드의 흐름이 동기적으로 실행이 되고 있기에 문제가 없어 보일 수는 있습니다. 하지만 현실 세상에서 커피를 만드는데는 시간이 걸리는걸 고려해서 pickUpTable을 관찰해서 변경사항이 생기면 출력을 하는 방식으로 수정을 해볼까요? 

> Hint) Property Observer

3. 전달인자 레이블의 남용

전달인자 레이블은 파라미터의 역할을 명확하게 표현하고 가독성을 높일 때 사용을 하곤 합니다. 하지만 현재 Hisop님은 파라미터의 기존 네이밍으로도 설명이 가능한 사항에서도 전달인자 레이블을 사용하는 경향이 있습니다. 그리고 전에도 말씀드렸듯이 전달인자 레이블은 전치사를 많이 사용하곤 합니다.

그렇다면 꼭 필요한 부분에만 사용을 할 수 있도록 수정해볼까요?

Commit:
Refactor: unwrap 네이밍 수정, pickUpTable 추가 방식 수정
네이밍을 unwrapedMenu에서 menuPrice로 변경
pickUpTable에 동일인이 주문한 메뉴가 있을경우 / 없을경우 구분하여 추가하도록 수정

Commit:
Refactor: pickUpTable에 변경사항 생기면 출력하도록 변경
고객이 커피를 주문할때 배열로 주문하도록 변경
pickuplist를 출력하는 메소드 추가

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -34,27 +34,32 @@ struct CoffeeShop {
+        }
+        else {
+            pickUpTable[name]?.append(coffee)
+        }
```
`pickUpTable[name] = [coffee]` 
name 이라는 키가 딕셔너리에 존재하고 있을 때 위 코드를 재실행하면 어떤 결과가 나올까요?

Hi-sop replied:
이미 존재하는 pickUpTable에 추가해준다는 의미로 사용했습니다.

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -12,55 +12,65 @@ struct Person {
-        guard let menuPrice = menu[coffee] else {
-            print("해당 메뉴는 존재하지 않습니다.")
-            return false
+    var pickUpTable: [String: [Coffee]] = [:] {
+        didSet {
+            printPickupList()
```
Property Observer 를 잘 활용해주셨네요 👍🏻
그렇다면 willSet 과 didSet 은 어떤 차이가 있을까요?

Hi-sop replied:
프로퍼티를 감시하고 있다가 변경 전 / 변경 후에 실행됩니다

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -12,55 +12,65 @@ struct Person {
-            return false
+    }
+        
+    func printPickupList() {
+        pickUpTable.forEach{ name, coffees in
+            let menu = coffees.map { $0.rawValue }.joined(separator: ", ")
+            print("\(name) 님이 주문하신 \(menu)(이/가) 준비되었습니다. 픽업대에서 가져가주세요.")
```
👍🏻👍🏻

juun97 commented on CodeStarterCamp_Week3/main.swift:
```diff
@@ -12,55 +12,65 @@ struct Person {
-    mutating func takeOrder(_ coffee: Coffee, orderedBy customer: Person) {
-        guard let menuPrice = menu[coffee] else {
-            return
+    func validateOrder(_ coffeeList: [Coffee], orderedBy customer: Person) -> (Bool, Int) {
+        var allPrice = 0
+        for coffee in coffeeList {
+            guard let menuPrice = menu[coffee] else {
+                print("해당 메뉴는 존재하지 않습니다.")
+                return (false, 0)
+            }
+            allPrice += menuPrice
+        }
+        guard customer.money >= allPrice || allPrice != 0 else {
+            print("잔액이 \(allPrice - customer.money)원만큼 모자랍니다.")
+            return (false, 0)
         }
```
주문의 유효성 검증을 한다는 함수같은데 Int 도 같이 배출 해주는 이유가 있을까요??

Hi-sop replied:
커피를 배열로 넘겨주게 되면서 총액을 계산하는 과정을 여러번 시행하기 싫어 계산한 allPrice를 넘겨주게 되었습니다.

Hi-sop replied:
```swift
let (validateMenu, price) = coffeeShop.validateOrder(coffeeList, orderedBy: self)
        if validateMenu {
            coffeeShop.takeOrder(coffeeList, orderedBy: self, allPrice: price)
            money -= price
        }
```
튜플로 받아 person의 프로퍼티를 수정하는데 썼습니다

juun97 reviewed (commented):
전체적으로 수정이 잘 이루어진 것 같습니다 👍🏻
남겨드린 코멘트 답변 만 주시면 바로 머지하도록 하겠습니다 😃
